packages/graphql-language-service/src/utils/getVariablesJSONSchema.ts

```diff
@@ -139,22 +140,48 @@ function getJSONSchemaFromGraphQLType(
   if ('defaultValue' in type && type.defaultValue !== undefined) {
     definition.default = type.defaultValue as JSONSchema4Type | undefined;
   }
+
   if (isEnumType(type)) {
-    definition.type = 'string';
     definition.enum = type.getValues().map(val => val.name);
+    if (!isNonNull) {
+      definition.enum.push(null);
+    }
   }
 
-  if (isScalarType(type) && scalarTypesMap[type.name]) {
-    definition.type = scalarTypesMap[type.name];
+  if (isScalarType(type)) {
+    // default scalars
+    if (scalarTypesMap[type.name]) {
+      if (isNonNull) {
+        definition.type = scalarTypesMap[type.name];
+      } else {
+        definition.type = [scalarTypesMap[type.name], 'null'];
+      }
+    } else {
+      // custom scalar can be of any type
+      // FIXME: allow custom scalars to be configured somewhere?
```

Comment:
these are determined from the custom scalar spec https://graphql.org/blog/2023-01-14-graphql-scalars/

so the mapping should be based on the provided scalars in the official repo

Reply:
Good point, we probably could implement validation for "official" specs.

But one might want custom or non-spec'ed scalars, so I'd still like a way to inject custom validation.
(example, [AWS AppSync scalars.](https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html)). 

Also, one might want to override the community specs (e.g. custom `DateTime` that does not follow the specs for some reason).

I'd be happy to work on that, but it's kinda out of the scope of this PR.

Reply:
indeed, i agree it's out of scope, just adding a meta-comment for your comment haha